---
title: Prediction of maladaptation from ecological and genomic data using genomic offsets
subtitle: Master's Thesis in Bioinformatics
format: 
  revealjs:
    embed-resources: true
    theme: [default, clean.scss]
    menu: false
    slide-number: true
    date-format: long
    logo: img/aulogo_uk_var1_black.png
    footer: "Bioinformatics Research Center at Aarhus University"
output-dir: slides
html-math-method:
  method: mathjax
  url: "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"
author:
  - name: Curro Campuzano Jim√©nez
    orcid: 0000-0001-8285-9318
    email: campuzanocurro@gmail.com
    affiliations: Bioinformatics Research Center at Aarhus University
date: last-modified
---

## 

::: callout-note
## Genomic offsets

A set of statistical tools that predict the maladaptation of populations to rapid environmental change based on genotypes $\times$environment association models
:::

![](img/publications.png){#fig-publi fig-align="center" width="100%"}

## Outline

1.  Overview of the Gain and collaborators (2023) model
2.  Methodological analysis with general simulations
    -   Comparison of different methods under different scenarios

    -   Identifying putatively adaptive loci

    -   Measuring uncertainty
3.  Case study: Mediterranean thyme

```{=html}
<!--
## 

::: {.callout-note icon="false"}
### Genomic offsets

A set of statistical methods that combine genomic and ecological data to predict maladaptation in the event of an instantaneous environmental change
:::

![Articles in Web of Science with keywords 'Genomic offset', 'Genetic offset', 'RONA' or 'Genomic vulnerability'](img/publications.svg){#fig-oldpaper fig-align="center" width="100%"}

::: notes
They are getting more attention. Combine ecological and genomic data.
:::

-->
```
## An explicit model of genomic offsets by Gain and collaborators

![](img/framework.png){#fig-model fig-align="center" width="100%"}

::: notes
1.  QTLs
2.  Polygenic score (additive effect size)
3.  Heterogeneous environment
4.  Gaussian fitness curve -\> shifted fitness
:::

## Gaussian stabilizing selection

$$
w(z| \mathbf{x}^*) =   \exp\left(\frac{-\left(z - z_{\text{opt}}(\mathbf x ^*) \right)^2}{2V_S}\right)
$$

## Fit a genotype $\times$ environment association model

![](img/phenotypes.png){#fig-gea width="100%" fig-align="center"}

We have to assume all individuals are within their adaptive optimum and we can measure the QTLs!

## With a bit of math rearranging ...

$$
G^2(\mathbf{x}, \mathbf{x}^*) = \frac{\left(\sum _l ^L \hat y_l(\mathbf x) - \hat y_l(\mathbf x^*)\right)^2}{L}
$$

Under Gaussian stabilizing selection we would find that a relationship between the genomic offset and shifted fitness:

$$
    \mathbb E[-\log (w(\mathbf{x}, \mathbf{x}^*)] \approx  \frac{a^2\mathcal G^2(\mathbf{x}, \mathbf{x}^*)}{2V_s}
$$

## Genomic offsets in a nutshell

1.  Sample locally optimal individuals and measure their genotypes and current environment
2.  Identify a set of putatively adaptive loci using hypothesis testing
3.  Fit a genotype $\times$ environment association statistical model
4.  Calculate genomic offset between current and shifted environment
5.  Rank individuals / populations

# Results {.r-fit-text background-color="#e64173"}

## Simulated data using SLiM

![](img/slim_simulation.gif){#fig-animation fig-align="center" width="100%"}

## Data

![](img/causal_empirical.svg){#fig-causalempirical fig-align="center" width="100%"}

## Different local and non-local adaptation scenarios

![](img/geometric_go2.png){fig-align="center" width="100%"}

## No differences between methods if using the same set of candidates

![](img/different_methods.svg){#fig-different fig-align="center" width="100%"}

## How to identify the putatively adaptive loci?

-   Hypothesis testing approach based on genotype$\times$ environment association model

-   Other options?

## Minimize false negatives!

![Weak and strong asymmetry refer to the difference in relative importance of the two adaptive phenotypes.](img/fp_versus_fn.svg){#fig-fpfn fig-align="center" width="100%"}

## What about bootstrapping to measure uncertainty?

![](img/boots.png){#fig-boots fig-align="center" width="100%"}

## What about bootstrapping to measure uncertainty?

![](img/boots2.png){#fig-boots2 fig-align="center" width="100%"}

## Representative case: Mediterranean thyme

![](img/ecotypes.svg){fig-align="center" width="100%"}

## Representative case: Mediterranean thyme

![](img/thyme.svg){fig-align="center" width="100%"}

## Modest predictive power of shifted fitness

![](img/thyme_shifted.svg){#fig-thymeshift fig-align="center" width="100%"}

## Future work

1.  Explicitly model two or more latent phenotypic trait
2.  Explore a larger space of hyperparameters in the simulations
3.  Consider different approaches for identifying putatively adaptive loci
4.  Further study the bootstrapping approach
5.  Extend thyme simulations
6.  Analyze real data!

## Take home message

- Genomic offsets are an effective approach if you have external evidence of populations being locally optimal

- Identifying putatively adaptive loci using hypothesis-testing improves performance, but not be too strict

-   Measuring the uncertainty of your estimates using bootstrapped ranked genomic offsets is a promising strategy

-   Simulate specific data to show the method could work in theory, or if it will fail (non-continuous phenotypes, migration ...)

# Thanks! {.r-fit-text background-color="#e64173"}

# Extra slides

## Most popular methods

Different methods rely on different genotype $\times$ environment association model and define slightly different metrics

1.  Geometric genomic offset
2.  Redundancy analysis genomic offset
3.  Risk of non-adaptedness (RONA)
4.  Gradient forest genomic offset

## Alternative genomic offsets

![](img/methods_all.png){#fig-alternative fig-align="center" width="100%"}

## Conceptual issues on genomic offsets

### Locally optimal versus locally adapted

-   Assumptions have been stated vaguely before
-   I argue that methods assume sampled individuals are within their adaptive optimum
-   Results of Gain and collaborators hold if latent phenotype has constant variance

## Conceptual issues in genomic offset framework

![](img/interpolation.svg){#fig-interpolation fig-align="center" width="100%"}

## Distribution of genomic offsets

![](img/hist_offsets.svg){#fig-dist fig-align="center" width="100%"}

## Different local and non-local adaptation scenarios

![](img/criteria.svg){#fig-criteria fig-align="center" width="100%"}

## Hypothesis-testing prevents spurious inference when phenotype has no genetic basis

![](img/random.svg){#fig-random fig-align="center" width="100%"}

Lind and collaborators found that randomly selected were as good as putatively adaptive loci.


## Selecting environmental variables

-   Model building and variable selection?
-   Prior knowledge?

## Genomic offsets are robust to uncorrelated environmental variables

![](img/uncorrelated.svg){#fig-uncor fig-align="center" width="100%"}

## Be cautious and look for potential confounded variables!

![](img/confounded.png){#fig-confoun fig-align="center" width="100%"}

## Spearman's correlation of bootstrapped genomic offsets

![](img/error_bars.svg){#fig-spearman fig-align="center" width="100%"}

## 

![](img/pseudocode.svg){#fig-code fig-align="center" width="100%"}

## A reasonably comprehensive Julia package

![](img/julia.png){#fig-jul fig-align="center" width="100%"}

## Runtimes

![](img/time.svg){#fig-time fig-align="center" width="100%"}

## Numerical issues

![](img/numerical_issues.svg){#fig-numerical fig-align="center" width="80%"}

## Near-zero predictive power of average future fitness

![](img/thyme_future.svg){#fig-thymefuture fig-align="center" width="100%"}
